android-core/src/main/kotlin/com.mparticle/JobSchedulerUtils.kt

```diff
@@ -0,0 +1,83 @@
+package com.mparticle
+
+import android.app.job.JobInfo
+import android.app.job.JobScheduler
+import android.content.ComponentName
+import android.content.Context
+import android.os.Build
+import androidx.annotation.RequiresApi
+import com.mparticle.internal.Logger
+import com.mparticle.uploadbatching.BatchUploadingJob
+
+const val UPLOAD_BATCH_JOB = 123
+
+fun Long.secondsToMillis(): Long = this * 1000
+fun Long.minutesToMillis(): Long = (this * 60).secondsToMillis()
+
+enum class SchedulingBatchingType { PERIODIC, ONE_SHOT; }
+
+fun cancelScheduledUploadBatchJob(context: Context) {
+    try {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            val jobScheduler = context.getJobScheduler()
+            val jobRunning = context.getScheduledJob(UPLOAD_BATCH_JOB)
+            jobRunning?.let { jobScheduler?.cancel(UPLOAD_BATCH_JOB) }
+        }
+    } catch (e: Exception) {
+    }
+}
+
+fun scheduleBatchUploading(
+    context: Context,
+    delayInSeconds: Long,
+    type: SchedulingBatchingType,
+    cancelPrevious: Boolean = true,
+    legacyAction: (delay: Long) -> Unit
+) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        try {
+            if (cancelPrevious) {
+                context.getJobScheduler()?.cancel(UPLOAD_BATCH_JOB)
+            }
+            val builder = JobInfo.Builder(
+                UPLOAD_BATCH_JOB,
+                ComponentName(context, BatchUploadingJob::class.java.name)
+            )
+            if (type == SchedulingBatchingType.PERIODIC) {
+                builder.setPeriodic(delayInSeconds.secondsToMillis())
+            } else if (type == SchedulingBatchingType.ONE_SHOT) {
+                builder.setMinimumLatency(delayInSeconds.secondsToMillis())
+            }
+            builder.apply {
+                setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
+            }
+            context.scheduleJob(builder.build())
+        } catch (e: Exception) {
+            Logger.error("Error while trying to use JobSceduler for batch upload")
```

Comment:
When using this I get spammed with error messages saying "Error while trying to use JobSceduler for batch upload" (including the typo..).

Can we check programmatically if the service is there rather than relying on an exception?

And - rather than a vague error, we should log more or a warning that includes the code that they need to add to their manifest, and let them know that as a result of this, we may not be able to upload in the background.